Modules/_asynciomodule.c

```diff
@@ -612,19 +613,25 @@ create_cancelled_error(PyObject *msg)
 }
 
 static void
-set_cancelled_error(PyObject *msg)
+future_set_cancelled_error(FutureObj *fut)
 {
-    PyObject *exc = create_cancelled_error(msg);
+    PyObject *exc = create_cancelled_error(fut->fut_cancel_msg);
     PyErr_SetObject(asyncio_CancelledError, exc);
     Py_DECREF(exc);
+
+    PyThreadState *tstate = _PyThreadState_GET();
+
+    _PyErr_StackItem *saved_exc_info = tstate->exc_info;
```

Comment:
Oh, this is fairly invasive. Ideally we shouldn't be doing something like this in non-core files.

Reply:
Can we perhaps make `_PyErr_ChainThreadState` a bit more versatile to handle this use case? Maybe add a second argument?

Reply:
@1st1 Thanks for the helpful comments, Yury.

> Can we perhaps make _PyErr_ChainThreadState a bit more versatile to handle this use case?

Sure, I'll propose something. This is something that had crossed my mind, too.